scalding-macros/src/test/scala/com/twitter/scalding/macros/MacroOrderingProperties.scala

```diff
@@ -0,0 +1,137 @@
+      anOption <- arb[Option[Int]]
+    } yield TestCC(aInt, aLong, anOption, aDouble)
+  }
+}
+case class TestCC(a: Int, b: Long, c: Option[Int], d: Double)
+
+class MacroOrderingProperties extends FunSuite with PropertyChecks with ShouldMatchers with LowerPriorityImplicit {
+
+  def serialize[T](t: T)(implicit orderedBuffer: OrderedBufferable[T]): ByteBuffer = {
+    val buf = ByteBuffer.allocate(128)
+    orderedBuffer.put(buf, t)
+    buf.position(0)
+    buf
+  }
+
+  def rt[T](t: T)(implicit orderedBuffer: OrderedBufferable[T]) = {
+    val buf = serialize[T](t)
+    orderedBuffer.get(buf).map(_._2).get
+  }
+
+  def rawCompare[T](a: T, b: T)(implicit obuf: OrderedBufferable[T]): Int = {
+    obuf.compareBinary(serialize(a), serialize(b)).unsafeToInt
+  }
+
+  def oBufCompare[T](a: T, b: T)(implicit obuf: OrderedBufferable[T]): Int = {
+    obuf.compare(a, b)
+  }
+
+  def clamp(i: Int): Int =
+    i match {
+      case x if x < 0 => -1
+      case x if x > 0 => 1
+      case x => 0
+    }
+  def check[T: Arbitrary](implicit ord: Ordering[T], obuf: OrderedBufferable[T]) = forAll { (a: T, b: T) =>
+    assert(oBufCompare(rt(a), a) == 0, "A should be equal to itself after an RT")
+    assert(oBufCompare(rt(b), b) == 0, "B should be equal to itself after an RT")
+    rawCompare(a, b) + rawCompare(b, a) shouldEqual 0
+    oBufCompare(a, b) + oBufCompare(b, a) shouldEqual 0
+
+    assert(oBufCompare(rt(a), rt(b)) === oBufCompare(a, b), "Comparing a and b with ordered bufferables compare after a serialization RT")
+
+    assert(rawCompare(a, b) === clamp(ord.compare(a, b)))
+
+    assert(oBufCompare(a, b) === clamp(ord.compare(a, b)))
+  }
+
+  test("Test out Int") {
+    implicit val localOrdering = Ordering.ordered[Int](identity)
+    check[Int]
+  }
+
+  test("Test out Long") {
+    implicit val localOrdering = Ordering.ordered[Long](identity)
+    check[Long]
+  }
+
+  test("Test out Short") {
+    implicit val localOrdering = Ordering.ordered[Short](identity)
+    check[Short]
+  }
+
+  test("Test out Float") {
+    implicit val localOrdering = Ordering.ordered[Float](identity)
+    check[Float]
+  }
+
+  test("Test out Double") {
+    implicit val localOrdering = Ordering.ordered[Double](identity)
+    check[Double]
+  }
+
+  test("Test out Char") {
+    implicit val localOrdering = Ordering.ordered[Char](identity)
+    check[Char]
+  }
+
+  test("Test out Byte") {
+    implicit val localOrdering = Ordering.ordered[Byte](identity)
+    check[Byte]
+  }
+
+  test("Test out Option[Int]") {
+    implicit val localOrdering = Ordering.Option(Ordering.Int)
+    check[Option[Int]]
+  }
+
+  test("Test out Option[Option[Int]]") {
+    implicit val localOrdering = Ordering.Option(Ordering.Option(Ordering.Int))
+    check[Option[Option[Int]]]
+  }
+
+  test("Test out TestCC") {
+    import TestCC._
+    check[TestCC](implicitly[Arbitrary[TestCC]], Ordering.by(t => TestCC.unapply(t)), implicitly)
+  }
+
+  test("Test out (Int, Int)") {
```

Comment:
what about a class that specifically exercises Product?

```scala
class MyData(override val _1: Int, override val _2: Option[Long]) extends Product2[Int, Option[Long]]
```
just to be sure the tuple case is not picking up case class version.